svi/auto_thermal_reformer/condition_num_plots.py

```diff
@@ -0,0 +1,168 @@
+    (0.92, 1587379.0, 58),
+    (0.93, 1657379.0, 52)
+]
+
+SUBSETS_IMPLICIT = [
+    (0.94, 1947379.0, 45),
+    (0.90, 1727379.0, 35),
+    (0.92, 1587379.0, 41),
+    (0.93, 1657379.0, 37)
+]
+
+SUBSETS_ALAMO = [
+    (0.94, 1947379.0, 29),
+    (0.90, 1727379.0, 26),
+    (0.92, 1587379.0, 32),
+    (0.93, 1657379.0, 27)
+]
+
+SUBSETS_NN = [
+    (0.94, 1947379.0, 53),
+    (0.90, 1727379.0, 48),
+    (0.92, 1587379.0, 47),
+    (0.93, 1657379.0, 44)
+]
+
+def calculate_condition_number(m):
+    nlp = PyomoNLP(m)
+    jac = nlp.evaluate_jacobian()
+    cond_num = np.linalg.cond(jac.toarray())
+    return cond_num
+
+def full(X=0.94, P=1947379.0, iters=300):
+    m = make_optimization_model(X, P)
+    solver = config.get_optimization_solver(iters=iters)
+    print(f"Solving sample with X={X}, P={P}")
+    results = solver.solve(m, tee=True)
+    cond_num = calculate_condition_number(m)
+    return cond_num
+
+def implicit(X=0.94, P=1947379.0, iters=300):
+    m = make_optimization_model(X, P)
+    add_external_function_libraries_to_environment(m)
+    m_implicit = make_implicit(m)
+    solver = config.get_optimization_solver(iters=iters)
+    print(f"Solving sample with X={X}, P={P}")
+    results = solver.solve(m_implicit, tee=True)
+    cond_num = calculate_condition_number(m)
+    return cond_num
+
+def alamo(X=0.94, P=1947379.0, iters=300):
+    m = create_alamo_instance(X, P, surrogate_fname=os.path.join(data_dir, DEFAULT_ALAMO_SURROGATE_FNAME))
+    initialize_alamo_atr_flowsheet(m)
+    m.fs.reformer_bypass.inlet.temperature.unfix()
+    m.fs.reformer_bypass.inlet.flow_mol.unfix()
+    solver = config.get_optimization_solver(iters=iters)
+    print(f"Solving sample with X={X}, P={P}")
+    results = solver.solve(m, tee=True)
+    cond_num = calculate_condition_number(m)
+    return cond_num
+
+def nn(X=0.94, P=1947379.0, iters=300):
+    m = create_nn_instance(X, P, surrogate_fname=os.path.join(data_dir, DEFAULT_NN_SURROGATE_FNAME))
+    initialize_nn_atr_flowsheet(m)
+    m.fs.reformer_bypass.inlet.temperature.unfix()
+    m.fs.reformer_bypass.inlet.flow_mol.unfix()
+    solver = config.get_optimization_solver(iters=iters)
+    print(f"Solving sample with X={X}, P={P}")
+    results = solver.solve(m, tee=True)
+    cond_num = calculate_condition_number(m)
+    return cond_num
+
+plt.figure(figsize=(12, 8))
+
+for i, (full_subset, implicit_subset, alamo_subset, nn_subset) in enumerate(zip(SUBSETS_FULLSPACE, 
+                                                                                SUBSETS_IMPLICIT, 
+                                                                                SUBSETS_ALAMO, 
+                                                                                SUBSETS_NN), 1):
+    plt.subplot(2, 2, i)
+    full_list = []
+    implicit_list = []
+    alamo_list = []
+    nn_list = []
+    for iters in range(1, full_subset[2] + 1):
+        full_list.append(full(X=full_subset[0], P=full_subset[1], iters=iters))
```

Comment:
We shouldn't need to re-run the optimization to compute condition number each iteration. We can use `ConditioningCallback` to make this easier:
```python
solver = config.get_optimization_solver()
from svi.cyipopt import ConditioningCallback
# Set an intermediate callback that calculates condition number each iteration
callback = ConditioningCallback()
solver.config.intermediate_callback = callback
solver.solve(model, tee=True)
full_list = callback.condition_numbers
```

Reply:
Yes! Thank you, I was modifying this script to include the ConditioningCallback, I'll push it soon. I did not include in `run_fullspace_sweep.py` because it took too long to run the 64 instances, but I'll include it in `condition_num_plots.py`.

Reply:
Just as personal reminder of our conversation, an example would be to plot iterations vs constraint residual of sum of mole fractions in reformer recuperator.